File: app/src/main/java/com/wadektech/eventshub/adapter/ConcertsAndTheatreAdapter.java
package com.wadektech.eventshub.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.wadektech.eventshub.R;
import com.wadektech.eventshub.models.Concerts;

import java.util.List;

public class ConcertsAndTheatreAdapter extends RecyclerView.Adapter<ConcertsAndTheatreAdapter.ViewHolder> {
    public List<Concerts> concertsList;
    public Context context ;
    public OnSingleConcertItemClicked singleConcertItemClicked ;

    public ConcertsAndTheatreAdapter(List<Concerts> concertsList, Context context, OnSingleConcertItemClicked singleConcertItemClicked) {
        this.concertsList = concertsList;
        this.context = context;
        this.singleConcertItemClicked = singleConcertItemClicked;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.events_list, parent, false);
        return new ViewHolder(view , singleConcertItemClicked);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Concerts concerts = concertsList.get(position);

        holder.date.setText(String.valueOf(concerts.getDate()));
        holder.entryFees.setText(String.valueOf(concerts.getEntryFees()));
        holder.eventTitle.setText(concerts.getTitle());
        holder.location.setText(concerts.getLocation());
        holder.shortDescription.setText(concerts.getShortDesc());

    }

    @Override
    public int getItemCount() {
        return concertsList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView eventTitle , shortDescription, date, location , entryFees ;
        OnSingleConcertItemClicked concertItemClicked ;
        public ViewHolder(@NonNull View itemView , OnSingleConcertItemClicked concertItemClicked) {
            super(itemView);
            eventTitle = itemView.findViewById(R.id.tv_event_title);
            shortDescription = itemView.findViewById(R.id.tv_title_desc);
            date = itemView.findViewById(R.id.tv_event_date);
            location = itemView.findViewById(R.id.tv_event_location);
            entryFees = itemView.findViewById(R.id.tv_event_fees);
            this.concertItemClicked = concertItemClicked;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            concertItemClicked.onSingleConcertItemClicked(getAdapterPosition());
        }
    }

    public interface OnSingleConcertItemClicked{
        void onSingleConcertItemClicked(int position) ;
    }
}

File: app/src/main/java/com/wadektech/eventshub/adapter/ConcertsAndTheatreAdapter.kt
package com.wadektech.eventshub.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.wadektech.eventshub.R
import com.wadektech.eventshub.databinding.ConcertsEventsListBinding

import com.wadektech.eventshub.models.Concerts

class ConcertsAndTheatreAdapter(var singleConcertItemClicked: OnSingleConcertItemClicked) :
        PagedListAdapter<Concerts, ConcertsAndTheatreAdapter.ViewHolder>(ConcertsDiffUtil()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder.from(parent)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val concerts = getItem(position)
        if (concerts != null){
            holder.bind(concerts , singleConcertItemClicked)
        }
    }

    class ViewHolder private constructor(val binding: ConcertsEventsListBinding) :
            RecyclerView.ViewHolder(binding.root) {

        fun bind(concerts: Concerts, singleConcertItemClicked: OnSingleConcertItemClicked){
            binding.tvEventTitle.text = concerts.title
            binding.tvTitleDesc.text = concerts.shortDesc
            binding.tvEventDate.text = concerts.date.toString()
            binding.tvEventLocation.text = concerts.location
            binding.tvEventFees.text = concerts.entryFees.toString()

            itemView.setOnClickListener {
                singleConcertItemClicked.onSingleConcertItemClicked(adapterPosition)
            }
        }
        companion object {
            fun from(parent: ViewGroup): ViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ConcertsEventsListBinding.inflate(layoutInflater, parent, false)
                return ViewHolder(binding)
            }
        }
    }

    interface OnSingleConcertItemClicked {
        fun onSingleConcertItemClicked(position: Int)
    }

    class ConcertsDiffUtil : DiffUtil.ItemCallback<Concerts>(){
        override fun areItemsTheSame(oldItem: Concerts, newItem: Concerts): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Concerts, newItem: Concerts): Boolean {
            return oldItem.id == newItem.id
        }
    }
}